include/soft_limiter.h

```diff
@@ -69,93 +77,79 @@ sequence of audio. For example:
   SoftLimiter<48> limiter("channel name", prescale);
 
 You can then repeatedly call:
-  const auto out_array = limiter.Apply(in_array, num_frames);
+  auto &&out = limiter.Apply(in, num_frames);
 
 Where:
- - in_array is a std::array<float, 48>
- - num_frames is some number of frames from zero up to 48.
- - out_array is a const reference to std::array<int16_t, 48>
+ - in is a std::vector<float>
+ - num_frames is some number of frames from zero up in.size()
+ - out is a a std::vector<int16_t> of samples (num_frames * 2).
 
-The limiter will either copy or scale-limit num_frames into the out_array.
+The limiter will either copy or scale-limit num_frames into the out vector.
 
-The PrintStats function will make mixer suggestions if the in_array samples were
-either significantly under or over the allowed bounds for the entire playback
-duration.
+The PrintStats function will make mixer-level suggestions if the 'in' samples
+were either significantly under or over the allowed bounds for the entire
+playback duration.
 */
 
-#include "dosbox.h"
-
-#include <algorithm>
-#include <array>
-#include <cassert>
-#include <limits>
-#include <cmath>
-#include <string>
-
-#include "logging.h"
-#include "mixer.h"
-#include "support.h"
-
-template <size_t array_frames>
 class SoftLimiter {
 public:
-	SoftLimiter(const std::string &name, const AudioFrame &scale);
-
 	// Prevent default object construction, copy, and assignment
 	SoftLimiter() = delete;
-	SoftLimiter(const SoftLimiter<array_frames> &) = delete;
-	SoftLimiter<array_frames> &operator=(const SoftLimiter<array_frames> &) = delete;
+	SoftLimiter(const SoftLimiter &) = delete;
+	SoftLimiter &operator=(const SoftLimiter &) = delete;
 
-	constexpr static size_t array_samples = array_frames * 2; // 2 for stereo
-	using in_array_t = std::array<float, array_samples>;
-	using out_array_t = std::array<int16_t, array_samples>;
+	SoftLimiter(const std::string &name,
+	            const AudioFrame &scale,
+	            const uint16_t max_frames);
 
-	out_array_t Apply(const in_array_t &in, uint16_t frames) noexcept;
+	using in_t = std::vector<float>;
+	using out_t = std::vector<int16_t>;
+	out_t Apply(const in_t &in, uint16_t req_frames) noexcept;
 	const AudioFrame &GetPeaks() const noexcept { return global_peaks; }
 	void PrintStats() const;
 	void Reset() noexcept;
 
 private:
-	using in_array_iterator_t = typename std::array<float, array_samples>::const_iterator;
-	using out_array_iterator_t = typename std::array<int16_t, array_samples>::iterator;
+	using in_iterator_t = typename std::vector<float>::const_iterator;
+	using out_iterator_t = typename std::vector<int16_t>::iterator;
```

Comment:
They're used heavily in many function interfaces; it would get pretty messy with `std::vector<float>::const_iterator` used instead.  Unless there's a better way?

Reply:
I think all ar almost all of functions using these in interfaces should not pass so many parameters list. You can make a simple struct (only in cpp file) and use it to pass input and output from `static` functions (instead of exposing everything through the header file). This would be easier to write, easier to read, easier to understand.

But it's probably out of scope of this PR at this point, so let's leave it be if such refactorization is too much work.

Reply:
yes, the parameter lists are big and I agree a struct to pass would be nicer; but it's it's tricky figuring out how many variables we'd include in the struct -- because the interfaces aren't the same.  

So the struct could be small and cover most use cases (but wouldn't reduce the interface sizes much), or the struct could be mostly all-encompassing, and then some portion of struct members just sit idle (or we pass nullptr's or zeros for them). 

I dunno. Then accessing the values piles on the struct name, so code thats like:

``` c++
while (in_pos != in_end) {
	const auto fitted = poly_a * (*in_pos * prescalar - poly_b) + poly_b;
	assert(fabsf(fitted) <= out_limits::max());
	*out_pos = static_cast<int16_t>(fitted);
	out_pos += 2;
	in_pos += 2;
}
```

becomes:

``` c++
while (vars.in_pos != vars.in_end) {
	const auto fitted = vars.poly_a * (*vars.in_pos * vars.prescalar - vars.poly_b) + vars.poly_b;
	assert(fabsf(fitted) <= out_limits::max());
	*vars.out_pos = static_cast<int16_t>(fitted);
	vars.out_pos += 2;
	vars.in_pos += 2;
}
```

But I fully agree it would collapse the big argument lists - which I would like to get rid of.

I original had all common arguments as members purely to reduce the interface sizes; however they didn't contribute to the end result and really were just temporaries: so I opted to be more struct with the code an really keep them as temporaries.  Maybe that was the wrong move (hmm).

If we can find a way to refactor without these compromises - I can do it now and keep that on a separate commit.  We also have the extensive unit tests that ensure correct operation.

Reply:
I think significantly improve it (without causing a bunch of duplicate code), it will take some hard thinking and likely restructuring.  I don't have the stomach for it - so best we move this to a later time.

With the unit tests proving hard functionality-correctness test, the low LOC count, and it being extremely stand-alone in nature (ie: there's nothing "DOSBox-specific" about the code), all of these lend themselves to any community developers wanting to help!